pkg/kubectl/cmd/rollout/rollout_history.go

```diff
@@ -0,0 +1,132 @@
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/util/errors"
+
+	"github.com/spf13/cobra"
+)
+
+// HistoryOptions is the start of the data required to perform the operation.  As new fields are added, add them here instead of
+// referencing the cmd.Flags()
+type HistoryOptions struct {
+	Filenames []string
+}
+
+const (
+	history_long    = `view previous rollout versions and configurations.`
+	history_example = `# View the rollout history of a deployment
+$ kubectl rollout history deployment/abc`
+)
+
+func NewCmdRolloutHistory(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &HistoryOptions{}
+
+	cmd := &cobra.Command{
+		Use:     "history (TYPE NAME | TYPE/NAME) [flags]",
+		Short:   "view rollout history",
+		Long:    history_long,
+		Example: history_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmdutil.CheckErr(RunHistory(f, cmd, out, args, options))
+		},
+	}
+
+	cmd.Flags().Int("version", 0, "See the details, including podTemplate of the version specified")
+	usage := "Filename, directory, or URL to a file identifying the resource to get from a server."
+	kubectl.AddJsonFilenameFlag(cmd, &options.Filenames, usage)
+	return cmd
+}
+
+func RunHistory(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string, options *HistoryOptions) error {
+	if len(args) == 0 && len(options.Filenames) == 0 {
+		return cmdutil.UsageError(cmd, "Required resource not specified.")
+	}
+	versionDetail := cmdutil.GetFlagInt(cmd, "version")
+
+	mapper, typer := f.Object()
+
+	cmdNamespace, enforceNamespace, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		ResourceTypeOrNameArgs(true, args...).
+		Latest().
+		Flatten().
+		Do()
+	err = r.Err()
+	if err != nil {
+		return err
+	}
+
+	infos, err := r.Infos()
+	if err != nil {
+		return err
+	}
+	info := infos[0]
+	mapping := info.ResourceMapping()
+	historyViewer, err := f.HistoryViewer(mapping)
+
+	errs := []error{}
+	for _, info := range infos {
+		versionToRC, err := historyViewer.History(info.Namespace, info.Name)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if versionDetail > 0 {
+			rc, ok := versionToRC[versionDetail]
+			if !ok {
+				return fmt.Errorf("unable to find version %d of deployment %s", versionDetail, info.Name)
+			}
+			fmt.Fprintf(out, "Deployment %s version %d:\n", info.Name, versionDetail)
+			s, err := kubectl.DescribePodTemplate(rc.Spec.Template)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			fmt.Fprintf(out, "%s\n", s)
+		} else {
+			// Sort the versionToRC map by version
+			var versions []int
+			for k := range versionToRC {
+				versions = append(versions, k)
+			}
+			sort.Ints(versions)
+
+			fmt.Fprintf(out, "VERSION\n")
```

Comment:
The initial version doesn't need to be too fancy, but how about displaying change-source?

It might also make sense to add a column for the image, since that will often be what changed.

In the future, we might also want to think about how to summarize what changed, perhaps by doing a strategic diff and categorizing different parts of the pod template: image (I), env (E), volumes (V), resources (R), scheduling constraints (S), other (O).